### PR TITLE
feat: :sparkles: add support for Thread resource

### PR DESCRIPTION
**Pull request by benshuk**

### CODE REVIEW

File: ai21/clients/common/assistant/thread.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import List
+
+from ai21.models.responses.thread_response import ThreadResponse, CreateMessagePayload
+
+
+class Thread(ABC):
```
Comment (Josephasafg) [inline ai21/clients/common/assistant/thread.py:9]: `Thread` or `Threads`?

File: ai21/clients/common/assistant/thread.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    _module_name = "threads"
+
+    @abstractmethod
+    def create(
+        self,
+        messages: List[CreateMessagePayload],
+        **kwargs,
+    ) -> ThreadResponse:
```
Comment (Josephasafg) [inline ai21/clients/common/assistant/thread.py:17]: Perhaps only call the response `Thread` and the class `Threads`?

File: ai21/clients/common/assistant/thread.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        pass
+
```
Comment (Josephasafg) [inline ai21/clients/common/assistant/thread.py:19]: Do we also want to support something like this? https://github.com/openai/openai-python/blob/main/src/openai/resources/beta/threads/messages.py#L32

where we also have `messages` method inside `Threads` class?

Comment (benshuk) [reply ai21/clients/common/assistant/thread.py:19]: yeah, that's the next PR :)

File: ai21/clients/common/assistant/thread.py (hunk 2 above)
Comment (benshuk) [reply ai21/clients/common/assistant/thread.py:17]: sounds good. I just went with the same naming convention of the previous PR...

File: ai21/clients/common/assistant/thread.py (hunk 1 above)
Comment (benshuk) [reply ai21/clients/common/assistant/thread.py:9]: sounds good. I just went with the same naming convention of the previous PR...

File: ai21/models/responses/thread_response.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+class MessageContentText(TypedDict):
+    type: Literal["text"]
+    text: str
+
+
+class CreateMessagePayload(TypedDict):
+    role: MessageRole
+    content: MessageContentText
+
```
Comment (Josephasafg) [inline ai21/models/responses/thread_response.py:20]: I think its worth separating the non response objects to be under `models` package as other objects are

File: ai21/models/responses/thread_response.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+MessageRole = Literal["assistant", "user"]
+
+
+class MessageContentText(TypedDict):
+    type: Literal["text"]
+    text: str
+
+
+class CreateMessagePayload(TypedDict):
```
Comment (Josephasafg) [inline ai21/models/responses/thread_response.py:17]: How about only `Message`?

Comment (benshuk) [reply ai21/models/responses/thread_response.py:17]: no :( 
when the user creates a message they can control only specific fields, while the `message` that returns from the API has some extra fields in it that aren't in the user's capability of changing.

Comment (Josephasafg) [reply ai21/models/responses/thread_response.py:17]: I was referring to the name of the object

**Review by Josephasafg — APPROVED** [review]

**Pull request merged**